src/libzfs_core/py_zfs_core_module.c

```diff
@@ -1,54 +1,686 @@
+	 */
+	if (error_name == NULL) {
+		error_name = PyUnicode_FromString(name);
+		if (error_name == NULL) {
+			Py_DECREF(pycode);
+			return;
+		}
+	}
+
+	v = PyObject_CallFunction(state->zc_exc, "s", msg);
+	if (v == NULL)
+		return;
+
+	/* This steals references to pycode and error_name */
+	attrs = Py_BuildValue(
+		"(NsNO)",
+		pycode,
+		msg,
+		error_name,
+		errors_tuple
+        );
+
+	if (attrs == NULL) {
+		Py_DECREF(v);
+		return;
+	}
+
+	err = PyObject_SetAttrString(v, "code", PyTuple_GetItem(attrs, 0));
+	if (err == -1) {
+		Py_DECREF(v);
+		Py_DECREF(attrs);
+		return;
+	}
+
+	err = PyObject_SetAttrString(v, "msg", PyTuple_GetItem(attrs, 1));
+	if (err == -1) {
+		Py_DECREF(v);
+		Py_DECREF(attrs);
+		return;
+	}
+
+	err = PyObject_SetAttrString(v, "name", PyTuple_GetItem(attrs, 2));
+	if (err == -1) {
+		Py_DECREF(v);
+		Py_DECREF(attrs);
+		return;
+	}
+
+	err = PyObject_SetAttrString(v, "errors", PyTuple_GetItem(attrs, 3));
+	if (err == -1) {
+		Py_DECREF(v);
+		Py_DECREF(attrs);
+		return;
+	}
+
+	PyErr_SetObject((PyObject *) Py_TYPE(v), v);
+	Py_DECREF(v);
+	Py_DECREF(attrs);
+	return;
+}
+
+static
+boolean_t py_snapname_to_nvpair(nvlist_t *list,
+				PyObject *item,
+				PyObject *dsname_set,
+				char *pool_name,
+				size_t pool_name_sz)
+{
+	const char *snap;
+	Py_ssize_t sz;
+	char dsname[ZFS_MAX_DATASET_NAME_LEN];
+	char *psep = NULL;
+	PyObject *py_dsname = NULL;
+	int err;
+
+	if (!PyUnicode_Check(item)) {
+		PyObject *repr = PyObject_Repr(item);
+		if (repr == NULL)
+			return B_FALSE;
+
+		PyErr_Format(PyExc_TypeError,
+			     "%R: item  is not a string",
+			     repr);
+
+		Py_DECREF(repr);
+		return B_FALSE;
+	}
+
+	snap = PyUnicode_AsUTF8AndSize(item, &sz);
+	if (snap == NULL)
+		return B_FALSE;
+
+	if (!zfs_name_valid(snap, ZFS_TYPE_SNAPSHOT)) {
+		PyErr_Format(PyExc_TypeError,
+			     "%s: not a valid snapshot name",
+			     snap);
+		return B_FALSE;
+	}
+
+	if (*pool_name == '\0') {
+		/* pool name hasn't been set yet */
+		strlcpy(pool_name, snap, pool_name_sz);
+		pool_name[strcspn(pool_name, "/@")] = '\0';
+	}
+
+	if (strncmp(snap, pool_name, strlen(pool_name)) != 0) {
+		PyErr_Format(PyExc_ValueError,
+			     "%s: snapshot is not within "
+			     "expected pool [%s]. All "
+			     "snapshots must reside in the "
+			     "same pool.",
+			     snap, pool_name);
+		return B_FALSE;
+	}
+
+	strlcpy(dsname, snap, sizeof(dsname));
+
+	psep = strstr(dsname, "@");
+	PYZFS_ASSERT(psep, "Missing snapshot separator");
+
+	*psep = '\0';
+
+	py_dsname = PyUnicode_FromString(dsname);
+	if (!py_dsname)
+		return B_FALSE;
+
+	if (PySet_Contains(dsname_set, py_dsname)) {
+		Py_DECREF(py_dsname);
+		PyErr_Format(PyExc_ValueError,
+			     "%s: multiple snapshots of the same "
+			     "dataset is not permitted.",
+			     snap);
+		return B_FALSE;
+	}
+
+	err = PySet_Add(dsname_set, py_dsname);
+	Py_DECREF(py_dsname);
+	if (err)
+		return B_FALSE;
+
+	fnvlist_add_boolean(list, snap);
```

Comment:
Do we need `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS` here for nvlist ops for lzc?

Reply:
We don't need, but we can add it.